01-ruby-fundamentals/keword-arguments.md

```diff
@@ -0,0 +1,175 @@
+  elsif type == :latte
+    price = 3.7
+  elsif type == :cappuccino
+    price = 3.2
+  else
+    puts "Invalid coffee type: #{type}"
+    return
+  end
+
+  # Modify for size
+  if size == :tall
+    # no change
+  elsif size == :grande
+    price *= 1.3
+  elsif size == :venti
+    price *= 1.6
+  else
+    puts "Invalid size: #{size}"
+  end
+
+  price
+end
+```
+
+The `coffee_price` method would be invoked as follows:
+
+```ruby
+price = coffee_price(:latte, :venti)
+puts "Your total is #{price} dollars"
+```
+
+Spend some time reading through the above code, then answer the following questions:
+- What's the _method signature_ of the `coffee_price` method?
+- What does `coffee_price` _return_?
+- Why is it useful to put this functionality in a method?
+- How does the method work? Does anything surprise you about it?
+- How would you make sure this method does the right thing?
+- What happens if your argument doesn't match what the method expects, like `coffee_price(:decaf, :grande)`?
+
+### Adding More Options
+
+Now that we've got the basics, it's time to add some of the options on our coffee menu.
+- Extra shots of espresso: $0.50 each
+- Served cold: $1.00 extra
+
+Let's update our code to match:
+
+```ruby
+def coffee_price(type, size, extra_shots, cold)
+  # Find the base price for this drink
+  if type == :drip
+    price = 1.5
+  elsif type == :latte
+    price = 3.7
+  elsif type == :cappuccino
+    price = 3.2
+  else
+    puts "Invalid coffee type: #{type}"
+    return
+  end
+
+  # Modify for size
+  if size == :tall
+    # no change
+  elsif size == :grande
+    price *= 1.3
+  elsif size == :venti
+    price *= 1.6
+  else
+    puts "Invalid size: #{size}"
+  end
+
+  # Charge for extra shots
+  price += extra_shots * 0.5
+
+  # Charge for a cold drink
+  if cold
+    price += 1
+  end
+
+  price
+end
+```
+
+The code is starting to get pretty complicated, but that makes sense because our coffee pricing logic is also complex.
+
+Our _method signature_ is also starting to get complex. Here are two examples of calling the new `coffee_price`:
+
+```ruby
+# Price a venti hot latte with 3 extra shots
+coffee_price(:latte, :venti, 3, false)
+
+# Price a tall iced drip
+coffee_price(:drip, :tall, 0, true)
+```
+
+There are a few problems with this:
+- The caller has to remember the order of the arguments
+- When reading the call, there's no way to tell which argument is which
+- The caller must provide values for all arguments, even for the "standard" way of doing things (i.e. no extra shots)
+
+All of these put the onus on the caller to do things right, which is just asking for trouble. **A well-written method is easy to use right and hard to use wrong** - by that standard, this is not good code.
+
+### Keyword Arguments
+
+One way to address these problems is with a technique called _keyword arguments_. Here's what they look like:
+
+```ruby
+# Define the method
+def coffee_price(type, size, extra_shots: 0, cold: false)
```

Comment:
I think it would be good to have one final gist that contains all of the code as it ends up so that there isn't any confusion about the code carryover from above